src/lockscreen.js

```diff
@@ -72,34 +81,49 @@ var Lockscreen = GObject.registerClass({
     }
 
     set lock(lock) {
-        if ('_lock' in this && this._lock === lock)
+        if (this._lock === lock)
             return;
         if (this._lockChangedId !== 0)
             this._manager.disconnect(this._lockChangedId);
         this._lockChangedId = this._manager.connect(
             `changed::${lock}`, this._updateLockStateWithLock.bind(this));
         this._lock = lock;
+
+        var path = `/app/share/lockscreens/${lock}`;
+        var dir = Gio.File.new_for_path(path);
+        this._assetsPath = dir.query_exists(null) ? path : '/app/share/lockscreens/default';
+
         this._updateLockStateWithLock();
     }
 
     _updateLockStateWithKey() {
         if (!this._key)
             return;
-        if (this._manager.hasKey(this._key)) {
-
-            /* TODO update the lock-screen graphics with can-be-opened image */
-        }
+        if (this._assetsPath && this._manager.hasKey(this._key))
+            this._playbin.setBackground(`file://${this._assetsPath}/has-key`);
     }
 
     _updateLockStateWithLock() {
-        if (!this._lock)
+        if (!this._lock) {
+            this._playbin.add_class('no-lock');
             return;
+        }
+
+        this._playbin.remove_class('no-lock');
+        if (this._assetsPath) {
+            var bg = this._manager.hasKey(this._key) ? 'has-key' : 'no-key';
+            this._playbin.setBackground(`file://${this._assetsPath}/${bg}`);
+        }
```

Comment:
Perhaps we can call `_updateLockStateWithKey()` here? It's unclear to me if we need to do it here in addition to what we do in `_updateLockStateWithKey()`.

Reply:
I'd also try to centralize every UI state change in `_updateUI()`, it's difficult to follow all the flow otherwise.